### PR TITLE
Upgrade LLVM to 3.9

### PR DESCRIPTION
**Pull request by ZackPierce**
Upgrade LLVM dependency to 3.9, as part of an effort to tackle some related aspects of https://github.com/smackers/smack/issues/219 .

Currently split into multiple commits to make review easier, though I can collapse per the contribution guidelines after review.

The build changes involve switching to sourcing the apt.llvm.org repo rather than llvm-apt.ecranbleu.org repo.  I had to manually remove references to llvm-apt.ecranblue.org in `/etc/apt/sources.list`. Users starting from scratch (or using vagrant) should not have to worry about that step.

The C-land changes are largely renames and function signature changes, or the consequences thereof.

The only truly incidental alterations are build-script additions to support linux-neon-16 in order to make my particular dev cycle turnaround faster. I can remove that bit if desired.

### CODE REVIEW

**Comment by ZackPierce** [comment]
Hmm, it looks like the Travis build [timed out on an unrelated request](https://travis-ci.org/smackers/smack/jobs/239608048) to Nuget. I'm uncertain what the policy on retries is. Perhaps or @zvonimir @michael-emmi know?

**Comment by michael-emmi** [comment]
I’ve just triggered a restart on the failing build; hopefully it will go through this time.

File: bin/build.sh [hunk 1]
```diff
@@ -221,21 +226,38 @@ then
     sudo zypper --non-interactive install ${DEPENDENCIES}
     ;;
 
+  linux-neon-1[6]*)
```
Comment (michael-emmi) [inline bin/build.sh]: Could this case possibly be merged with the following `linux-ubuntu` case? The only differences seem to be that `xential` replaces `trusty` in the LLVM apt repo URL. Merging might be achieved by adding a release-name table mapping `neon` to `xenial` and `ubuntu` to `trusty`.

File: bin/build.sh [hunk 2]
```diff
@@ -162,7 +162,12 @@ linux-opensuse*)
 
 linux-ubuntu-1[46]*)
   Z3_DOWNLOAD_LINK="https://github.com/Z3Prover/z3/releases/download/z3-4.4.1/z3-4.4.1-x64-ubuntu-14.04.zip"
-  DEPENDENCIES+=" clang-3.8 llvm-3.8 mono-complete libz-dev libedit-dev"
+  DEPENDENCIES+=" clang-3.9 llvm-3.9 mono-complete libz-dev libedit-dev"
+  ;;
+
+linux-neon-1[6]*)
```
Comment (michael-emmi) [inline bin/build.sh]: Could this case possibly be merged with the previous `linux-ubuntu` case?

**Comment by michael-emmi** [comment]
Thanks for the contribution @ZackPierce !

We will have to discuss when to merge this one with @zvonimir since previously we had been upgrading our LLVM compatibility according to some schedule which escapes me at the moment. In the worst case, we may have to maintain two separate branches for each LLVM version we want to support at any given time, e.g., 3.8 and 3.9; not sure if there’s any good reason to keep 3.8 around.

Your thoughts @zvonimir ?

One further technical comment/question: couldn’t we make the version-number symbolic in the `bin/build.sh` script? E.g., instead of changing every single instance of `xxxx-3.8` to `xxxx-3.9` we might have just written `xxxx-${LLVM_VERSION}` and updated the single `LLVM_VERSION` variable from `3.8` to `3.9`. Wouldn’t it be more reasonable to do this going forward?

**Comment by zvonimir** [comment]
Indeed, thanks for the contribution @ZackPierce, especially since I have been in charge of updating LLVM versions in the past and it is not the most exciting task :).
I am traveling now and I won't have time to look at this in more detail until this coming Friday or so.
I think there is no need to keep LLVM 3.8 version around. We decided at some point not to provide strong backwards compatibility when it comes to LLVM. So our strategy is to stay around a year or so behind LLVM so that we also get their bug-patch release as well before we bump up our LLVM version.
I like @michael-emmi's suggestion to make the LLVM version in our build script symbolic. So maybe you could try to do that update?

**Comment by ZackPierce** [comment]
Sure, I'll give those modifications a shot. Thanks for the speedy feedback.

**Comment by ZackPierce** [comment]
Alright, now the latest commit collapses the the ubuntu/neon branches, detects ubuntu codenames (`trusty` vs `xenial`) and adds LLVM version parameterization.

**Comment by Jinxxed** [comment]
Hi,

exactly 1 day before you submitted this pull request, I tried to make SMACK compile with LLVM 4.0 myself.
This seems to work pretty well, but as soon as I have a printf in the code somewhere, I get a Segmentation Fault, since IntType in LLVM is not a DerivedType anymore. I just checked and realized this is also the case for LLVM 3.9.
May I ask, how did you solve this issue?

Thank you!

**Comment by Jinxxed** [comment]
To be more precise, I the Segfault happens in lib/smack/SmackRep.cpp:687, with a line like
```c
printf("%d %d", a, b);
```
in the source code

**Comment by michael-emmi** [comment]
Shall we merge this one?

**Comment by zvonimir** [comment]
Before merging such potentially significant updates, I like to run the whole SV-COMP set so that we can catch unforeseen problems that our regressions miss, including performance issues. We still did not get a chance to do this since we've been working on testing other updates, and hence we do not have a stable baseline for comparison.

**Comment by zvonimir** [comment]
@ZackPierce : Sorry about the delay. We are now ready to thoroughly test this on the whole of SV-COMP benchmark set. Could you please rebase this to the latest develop and also fix these conflicts? Thanks!

**Comment by ZackPierce** [comment]
Roger that, will give it a go.

**Comment by ZackPierce** [comment]
Rebased, resolved conflicts, and added in an additional version-mismatch fix that was required due to the upstream work on develop (re: `getIntrinsicInstrCost` API changes).

**Comment by zvonimir** [comment]
Thanks @ZackPierce!
@shaobo-he : Could you please run this on the whole of SV-COMP? Thx!

**Comment by shaobo-he** [comment]
@zvonimir Yes, I am working on it.

**Pull request merged**